### ISSUE TITLE
oc new-build: can't figure out how to specify image stream to overwrite FROM

### ISSUE THREAD

**Issue by stevekuznetsov** [openshift/origin#17675]
I'm trying to create a `Build` that will take in source from one image and use another image as the build FROM. I've got this:
```
$ oc new-build --source-image pr-21-257717b093c5e0b975dd6042:src                                        \
               --source-image-path /data/src/github.com/openshift/image-registry/images/dockerregistry/ \
               --image-stream pr-21-257717b093c5e0b975dd6042:origin-base                                \
               --to pr-21-257717b093c5e0b975dd6042:origin-docker-registry                               \
               --output json                                                                            \
               --dry-run
error: the image match "ci-openshift-ansible/pr-21-257717b093c5e0b975dd6042:origin-base" for source repository "pr-21-257717b093c5e0b975dd6042:src" does not appear to be a source-to-image builder.

- to attempt to use this image as a source builder, pass "--strategy=source"
- to use it as a base image for a Docker build, pass "--strategy=docker"
```

So I add the `--strategy=docker` and I get:

```
$ oc new-build --source-image pr-21-257717b093c5e0b975dd6042:src                                        \
               --source-image-path /data/src/github.com/openshift/image-registry/images/dockerregistry/ \
               --image-stream pr-21-257717b093c5e0b975dd6042:origin-base                                \
               --strategy docker                                                                        \
               --to pr-21-257717b093c5e0b975dd6042:origin-docker-registry                               \
               --output json                                                                            \
               --dry-run
error: --strategy is specified and none of the arguments provided could be classified as a source code location
```

This seems wonky. 

/kind bug
/cc @bparees @csrwng

**Comment by bparees** [comment]
i'm not sure it's possible to use new-build to create a buildconfig w/ just image inputs, which it sounds like is what you're trying to do.

and if that's the case, this is "won't fix" because there's only so much flexibility we're going to put into oc new-build.  At some point you have to just sit down and write the buildconfig yaml yourself because the tool isn't going to do it for you.

**Comment by bparees** [comment]
also what kind of build are you actually trying to do?  a docker build or an s2i build?  if it's a docker build, where is your dockerfile supposed to come from?

**Comment by stevekuznetsov** [comment]
This is the type of build I would like to create:

```yaml
apiVersion: v1
kind: Build
metadata:
  labels:
    build: "${BUILD_NAME}"
    runid: "${RUN_ID}"
    job-id: "${JOB_ID}"
  generateName: "origin-pod-${BUILD_NUMBER}-${BUILD_NAME}-"
spec:
  output:
    pushSecret:
      name: "${IMAGE_SECRET_NAME}"
    to:
      kind: ImageStreamTag
      name: "${BUILD_NAME}:origin-pod"
  runPolicy: Serial
  serviceAccount: builder
  source:
    contextDir: "pod"
    images:
    - from:
        kind: ImageStreamTag
        name: "${BUILD_NAME}:binaries"
      pullSecret:
        name: "${IMAGE_SECRET_NAME}"
      paths:
      - destinationDir: "."
        sourcePath: "/go/src/github.com/openshift/origin/images/pod"
    type: Image
  strategy:
    dockerStrategy:
      pullSecret:
        name: "${IMAGE_SECRET_NAME}"
      dockerfilePath: Dockerfile
      from:
        kind: ImageStreamTag
        name: "${BUILD_NAME}:repo-base"
    type: Docker 
```

**Comment by stevekuznetsov** [comment]
So for that example I was hoping 

```
$ oc new-build --source-image is:binaries --source-path ....pod --imagestream is:repo-base --strategy docker
```

Would get me what I want

**Comment by stevekuznetsov** [comment]
FWIW I would also love to keep using `oc` as my YAML generator.

**Comment by bparees** [comment]
so there's a dockerfile inside is:src/path ?

**Comment by stevekuznetsov** [comment]
Yes

**Comment by bparees** [comment]
you could try passing --binary
I forget if we let you launch a binary build w/o actually supplying binary content.

**Comment by bparees** [comment]
(along with --strategy=docker)

**Comment by stevekuznetsov** [comment]
Yep, that works!

**Comment by stevekuznetsov** [comment]
/close

This was a user error (but maybe the message could suggest something -- the first error was 100000x amazing since it suggested `--strategy=docker`).

**Comment by stevekuznetsov** [comment]
/reopen

Wait no with `--binary --strategy docker` I get:

```json
                "strategy": {
                    "type": "Source",
                    "sourceStrategy": {
                        "from": {
                            "kind": "ImageStreamTag",
                            "namespace": "ci-openshift-ansible",
                            "name": "pr-21-257717b093c5e0b975dd6042:origin-base"
                        }
                    }
                },
```

I was too eager.

**Comment by bparees** [comment]
well that's certainly not right.

**Comment by bparees** [comment]
Seems like there are two bugs:

1) --binary is ignoring the --strategy=docker flag
2) providing an image input should count for purposes of supplying "source" (you shouldn't have to pass --binary to trick oc new-build into creating the build w/o providing an explicit source repo)

**Comment by coreydaley** [comment]
@bparees 
If you build using this PR and run the command leaving off the --binary, it still comes up as sourceStrategy in the BuildConfig.  If you specify `--strategy docker` should it override the build strategy no matter what?

Test Command:
`oc new-build --source-image centos:latest --source-image-path /tmp --image-stream httpd --to test:testing --strategy=docker -o json --dry-run`

Strategy section of BuildConfig:
```
"strategy": {
                    "type": "Source",
                    "sourceStrategy": {
                        "from": {
                            "kind": "ImageStreamTag",
                            "namespace": "openshift",
                            "name": "httpd:2.4"
                        }
                    }
                },

```

**Comment by bparees** [comment]
> If you specify --strategy docker should it override the build strategy no matter what?

yes. and in particular, "oc new-build --binary --strategy=docker foo" should result in a binary build w/ strategy docker.  

--strategy should override any auto-detection logic we have that attempts to guess at the build strategy.

**Comment by coreydaley** [comment]
> yes. and in particular, "oc new-build --binary --strategy=docker foo" should result in a binary build w/ strategy docker.

The example command given above (using httpd instead of foo) results in a correct build config.

```
source:
    binary: {}
    type: Binary
  strategy:
    dockerStrategy:
      from:
        kind: ImageStreamTag
        name: httpd:2.4
        namespace: openshift
    type: Docker

```

**Comment by bparees** [comment]
well whatever command @stevekuznetsov ran didn't, see his comment:
https://github.com/openshift/origin/issues/17675#issuecomment-350136307

he says he passed --binary and --strategy docker and got a source strategy build.

also your example for `oc new-build --source-image centos:latest --source-image-path /tmp --image-stream httpd --to test:testing --strategy=docker -o json --dry-run` shows the wrong strategy type, so there's definitely at least some cases where it's not working correctly.

I suspect @stevekuznetsov ran something like `oc new-build --strategy dockery --binary --source-image foo --source-image-path bar mybuild`

so basically --binary alone is fine, but once you also supply another source input along with --binary, it falls into source strategy despite the "--strategy=docker" arg.

**Comment by mrostamii** [comment]
I've fixed it after I save the git credential on the machine. It has been getting this error on GitLab CI and the local bash script.
```
git config --global credential.helper cache
```
The root cause of that is after you run `oc new-app ...` or `oc new-build ...` it's necessary to access to the git on the local machine.
If you do it on a script or a pipeline, it has skipped then gives an error like that.
